### PR TITLE
feat(ca-metrics): switch to new metrics (client events)

### PR DESCRIPTION
**Pull request by torgeadelin**
# COMPLETES #< INSERT LINK TO ISSUE >

## This pull request addresses

- switch from old postEvent to submitClientEvent
- it only includes the switching + unit tests
- it doesn't remove old code, that will be done separately
- it also includes some bug fixing (some events were not making sense or missing fields).

## by making the following changes

- refactor

### Change Type

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation update
- [ ] Tooling change
- [x] Internal code refactor

## The following scenarios where tested

-unit tests

### I certified that

- [x] I have read and followed [contributing guidelines](https://github.com/webex/webex-js-sdk/blob/master/CONTRIBUTING.md#submitting-a-pull-request)
- [x] I discussed changes with code owners prior to submitting this pull request

- [x] I have not skipped any automated checks
- [x] All existing and new tests passed
- [ ] I have updated the documentation accordingly

---

Make sure to have followed the [contributing guidelines](https://github.com/webex/webex-js-sdk/blob/master/CONTRIBUTING.md#submitting-a-pull-request) before submitting.

### CODE REVIEW

File: packages/@webex/plugin-meetings/src/meeting/index.ts [hunk 1]
```diff
@@ -3844,10 +3868,10 @@ export default class Meeting extends StatelessWebexPlugin {
...
         .then(() => {
           MeetingUtil.handleAudioLogging(this.mediaProperties.audioTrack);
-          Metrics.postEvent({
-            event: eventType.MUTED,
-            meeting: this,
-            data: {trigger: trigger.USER_INTERACTION, mediaType: MetricsMediaType.AUDIO},
+          NewMetrics.submitClientEvent({
+            name: 'client.muted',
+            payload: {trigger: 'user-interaction', mediaType: 'audio'},
+            options: {meetingId: this.id},
```
Comment (marcin-bazyl) [inline packages/@webex/plugin-meetings/src/meeting/index.ts:3874]: just a thought: meetingId for all these events seems like something that's mandatory, so not sure if it should be in "options" - options sound like things that are "optional". Shouldn't meeting id be in payload.identifiers?

File: packages/@webex/plugin-meetings/src/meeting/index.ts [hunk 2]
```diff
@@ -5310,14 +5330,15 @@ export default class Meeting extends StatelessWebexPlugin {
...
-            errors: [
-              Metrics.generateErrorPayload(2004, false, MetricsError.name.MEDIA_ENGINE, undefined),
-            ],
+        NewMetrics.submitClientEvent({
+          name: 'client.ice.end',
+          payload: {canProceed: false},
+          options: {
+            meetingId: this.id,
+            error: {
+              // TODO: adapt this code to work in separate JIRA
```
Comment (marcin-bazyl) [inline packages/@webex/plugin-meetings/src/meeting/index.ts:5339]: does the jira exist? can you create it so that it doesn't get forgotten?

File: packages/@webex/plugin-meetings/test/unit/spec/breakouts/events.ts [hunk 3]
```diff
@@ -30,17 +29,19 @@ describe('plugin-meetings', () => {
...
-            breakoutMoveId: 'breakoutMoveId',
-            breakoutSessionId: 'sessionId',
-            breakoutGroupId: 'groupId',
+        breakoutEvent.postMoveCallAnalyzer('client.breakout-session.join.response', eventInfo);
+        assert.calledWithMatch(NewMetrics.submitClientEvent, {
+          name: 'client.breakout-session.join.response',
+          payload: {
+            identifiers: {
+              breakoutMoveId: 'breakoutMoveId',
+              breakoutSessionId: undefined,
```
Comment (marcin-bazyl) [inline packages/@webex/plugin-meetings/test/unit/spec/breakouts/events.ts:40]: why undefined values here? these are breakout unit tests so this doesn't feel right

File: packages/@webex/plugin-meetings/test/unit/spec/locus-info/index.js [hunk 4]
```diff
@@ -1900,5 +1906,113 @@ describe('plugin-meetings', () => {
         );
       });
     });
+
+    describe('#isMeetingActive', () => {
+      it('sends client event correctly for state = inactive', () => {
```
Comment (marcin-bazyl) [inline packages/@webex/plugin-meetings/test/unit/spec/locus-info/index.js:1911]: nitpick: feels like these tests are all very similar and could be parametrized, but they're also quite short anyway so feel free to ignore this

File: packages/@webex/plugin-meetings/src/meeting/util.ts [hunk 5]
```diff
@@ -55,7 +55,10 @@ const MeetingUtil = {
       );
     }
 
-    Metrics.postEvent({event: eventType.MEDIA_REQUEST, meeting});
+    NewMetrics.submitClientEvent({
```
Comment (marcin-bazyl) [inline packages/@webex/plugin-meetings/src/meeting/util.ts:58]: we should probably move the code for sending 'client.locus.media.request' and 'client.locus.media.response' into LocusMediaRequest.sendHttpRequest() function because that's when the /media request is actually being sent out (we can do it in another PR)

File: packages/@webex/plugin-meetings/src/roap/request.ts [hunk 6]
```diff
@@ -96,7 +97,12 @@ export default class RoapRequest extends StatelessWebexPlugin {
       `Roap:request#sendRoap --> ${locusSelfUrl} \n ${roapMessage.messageType} \n seq:${roapMessage.seq}`
     );
 
-    Metrics.postEvent({event: eventType.MEDIA_REQUEST, meetingId});
+    NewMetrics.submitClientEvent({
```
Comment (marcin-bazyl) [inline packages/@webex/plugin-meetings/src/roap/request.ts:100]: same like my other comment - we should move this into LocusMediaRequest.sendHttpRequest()

File: packages/@webex/plugin-meetings/test/unit/spec/meetings/index.js [hunk 7]
```diff
@@ -827,6 +830,38 @@ describe('plugin-meetings', () => {
               },
             });
           });
+
+          //TODO: can't seem to make this pass
```
Comment (marcin-bazyl) [inline packages/@webex/plugin-meetings/test/unit/spec/meetings/index.js:834]: This test is failing because the assert is checked before the finally() code is executed. It's because `handleLocusEvent()` doesn't return the promise from the call to `this.create()`

File: packages/@webex/plugin-meetings/src/locus-info/index.ts [hunk 8]
```diff
@@ -430,10 +428,13 @@ export default class LocusInfo extends EventsScope {
         LoggerProxy.logger.warn(
           'Locus-info:index#isMeetingActive --> Call Ended, locus state is inactive.'
         );
-        Metrics.postEvent({
-          event: eventType.REMOTE_ENDED,
-          meetingId: this.meetingId,
+        NewMetrics.submitClientEvent({
+          name: 'client.call.remote-ended',
```
Comment (marcin-bazyl) [inline packages/@webex/plugin-meetings/src/locus-info/index.ts:432]: we should also remove all the const values from eventType defined in src/metrics/config.ts (like REMOTE_ENDED in this case), because they're not used anymore

File: packages/@webex/plugin-meetings/src/breakouts/events.ts [hunk 9]
```diff
@@ -1,37 +1,40 @@
...
-    event: events,
-    meetingId: eventInfo.meeting.id,
-    data: {
-      breakoutMoveId: eventInfo.breakoutMoveId,
-      breakoutSessionId: eventInfo?.currentSession?.sessionId,
-      breakoutGroupId: eventInfo?.currentSession?.groupId,
-    },
-  });
+const breakoutEvent: {
+  onBreakoutMoveRequest: any;
```
[the remark this answers is not in the document]
Comment (torgeadelin) [reply packages/@webex/plugin-meetings/src/breakouts/events.ts:5]: i mean i still don't know what eventInfo holds, so replacing it with (eventInfo: any) => void; doesn't provide too much value either, but added it anyways.

File: packages/@webex/plugin-meetings/src/breakouts/events.ts [hunk 10]
```diff
@@ -1,37 +1,40 @@
...
+  onBreakoutMoveRequest: (eventInfo) => {
+    breakoutEvent.postMoveCallAnalyzer('client.breakout-session.move.request', eventInfo);
+  },
+  onBreakoutMoveResponse: (eventInfo) => {
+    breakoutEvent.postMoveCallAnalyzer('client.breakout-session.move.response', eventInfo);
+  },
+  onBreakoutJoinResponse: (eventInfo) => {
+    breakoutEvent.postMoveCallAnalyzer('client.breakout-session.join.response', eventInfo);
+  },
+  postMoveCallAnalyzer: (event: ClientEvent['name'], eventInfo: any) => {
```
[the remark this answers is not in the document]
Comment (torgeadelin) [reply packages/@webex/plugin-meetings/src/breakouts/events.ts:19]: I don't see any type created in the breakouts for it.. I'll leave it as it is.

File: packages/@webex/plugin-meetings/src/meeting/index.ts (hunk 1 above)
Comment (torgeadelin) [reply packages/@webex/plugin-meetings/src/meeting/index.ts:3874]: The `payload` object contains "true" payload for the event, and by "true" I mean it follows the expected type/structure for the client event. It is not all of the payload tho, some of it is computed inside the new metrics class and we merge them together to form the final payload. The reason why meetingId is needed here is in order to compute the extra payload inside the class(locus related identifiers, correlationId, etc).

There will be events which won't require meetingId (as they will happen before the meeting exists), which is why it's optional.

File: packages/@webex/plugin-meetings/src/meeting/index.ts (hunk 2 above)
Comment (torgeadelin) [reply packages/@webex/plugin-meetings/src/meeting/index.ts:5339]: yes, https://jira-eng-gpk2.cisco.com/jira/browse/SPARK-435138

File: packages/@webex/plugin-meetings/test/unit/spec/breakouts/events.ts (hunk 3 above)
Comment (torgeadelin) [reply packages/@webex/plugin-meetings/test/unit/spec/breakouts/events.ts:40]: hmm, this is an actual bug! Thanks for catching it!
The previous implementation was broke, the parameters were not grabbed correctly from the eventInfo inside the source code.

Comment (torgeadelin) [reply packages/@webex/plugin-meetings/test/unit/spec/breakouts/events.ts:40]: fixed now

File: packages/@webex/plugin-meetings/src/meeting/util.ts (hunk 5 above)
Comment (torgeadelin) [reply packages/@webex/plugin-meetings/src/meeting/util.ts:58]: Have created a jira for this: https://jira-eng-gpk2.cisco.com/jira/browse/SPARK-440552

File: packages/@webex/plugin-meetings/src/roap/request.ts (hunk 6 above)
Comment (torgeadelin) [reply packages/@webex/plugin-meetings/src/roap/request.ts:100]: https://jira-eng-gpk2.cisco.com/jira/browse/SPARK-440552

File: packages/@webex/plugin-meetings/test/unit/spec/meetings/index.js (hunk 7 above)
Comment (torgeadelin) [reply packages/@webex/plugin-meetings/test/unit/spec/meetings/index.js:834]: how can I test this then?

File: packages/@webex/plugin-meetings/src/locus-info/index.ts (hunk 8 above)
Comment (torgeadelin) [reply packages/@webex/plugin-meetings/src/locus-info/index.ts:432]: i have a separate JIRA to do cleanup

Comment (torgeadelin) [reply packages/@webex/plugin-meetings/src/locus-info/index.ts:432]: https://jira-eng-gpk2.cisco.com/jira/browse/SPARK-440534

File: packages/@webex/plugin-meetings/src/breakouts/events.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (marcin-bazyl) [reply packages/@webex/plugin-meetings/src/breakouts/events.ts:19]: maybe ask Marmot team to add the types - they should know them

File: packages/@webex/plugin-meetings/test/unit/spec/meetings/index.js (hunk 7 above)
Comment (marcin-bazyl) [reply packages/@webex/plugin-meetings/test/unit/spec/meetings/index.js:834]: you could change `handleLocusEvent()` to return a promise - the production code ignores its return value anyway, but just adding `await testUtils.flushPromises();` in the test before the assert also fixes the test and doesn't affect the production code, so probably better to just do that

**Review by marcin-bazyl — APPROVED** [review]

**Review by mkesavan13 — APPROVED** [review]
Looks good to me. Just a few questions out of curiosity - 

- In the new metrics, why are we sending only the meeting Id instead of the whole meeting object? Won't it be less information on the dashboard for us?
- Why are we removing all the constants and replacing them with plain strings?

**Comment by torgeadelin** [comment]
@mkesavan13 
- meetingId is used to gather more information about the event inside the new metrics classes, it is not send over. Previously you would pass in either the meetingId or the whole meeting object. (but again, they were only used to grab other information about that specific event). that logic still applies, but the only difference is that im always passing down the id, and grab the meeting object internally.
- the new metrics classes are strongly typed, and therefore we don't need to store additional constants.

**Comment by mkesavan13** [comment]
@torgeadelin That makes sense. Thank you.
